compiler/rustc_lint/src/late.rs

```diff
@@ -380,11 +395,25 @@ fn late_lint_crate<'tcx, T: LateLintPass<'tcx> + 'tcx>(tcx: TyCtxt<'tcx>, builti
         only_module: false,
     };
 
+    // Note: `passes` is often empty. In that case, it's faster to run
+    // `builtin_lints` directly rather than bundling it up into the
+    // `RuntimeCombinedLateLintPass`.
     let mut passes: Vec<_> =
         unerased_lint_store(tcx).late_passes.iter().map(|mk_pass| (mk_pass)(tcx)).collect();
-    passes.push(Box::new(builtin_lints));
-    let pass = RuntimeCombinedLateLintPass { passes: &mut passes[..] };
+    if passes.is_empty() {
+        late_lint_crate_inner(tcx, context, builtin_lints);
+    } else {
+        passes.push(Box::new(builtin_lints));
+        let pass = RuntimeCombinedLateLintPass { passes: &mut passes[..] };
+        late_lint_crate_inner(tcx, context, pass);
+    }
```

Comment:
This is the important bit: there are two different calls to `late_lint_crate_inner`. The builtins-only one is faster than the `passes` one, even if `passes` has just one pass, because it doesn't involve slice iteration for each `check_foo` call.

Another thing worth mentioning: this new structure is better than the old structure when `passes` is non-empty. The original structure involved two AST traversals, roughly:
```
late_lint_crate_inner(builtin_lints);
if !passes.is_empty() {
    late_lint_crate_inner(passes);
}
```
The new structure always involves one AST traversal:
```
if !passes.is_empty() {
    late_lint_crate_inner(builtin_lints);
} else {
    late_lint_crate_inner(bulitin_lints + passes);
}
```
So this should make bootstrapping and clippy slightly faster, though the difference may not be significant in practice.